klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,165 @@
+        """
+        Attempts to register `directory` with the klaus app as a git repository.
+        If `directory` is not a git repo, this method does nothing.
+        """
+        try:
+            repo = FancyRepo(directory)
+        except dulwich.errors.NotGitRepository:
+            logger.debug("Dulwich says '%s' is not a git repository", directory)
+            return
+
+        matching = filter(lambda existing_repo: existing_repo.name == repo.name, self.app.repos)
+        if matching:
+             logger.warning("skipping repo with duplicate name:%s", repo)
+             return
+        logger.info("Adding repository, '%s' at '%s'", repo.name, directory)
+        self.app.repos.append(repo)
+
+    def on_created(self, event):
+        # We're only interested in directories; A git repo is always a directory
+        if type(event) != DirCreatedEvent:
+            return
+        logger.debug("Received directory creation event for '%s'", event.src_path)
+        self._add_dir(event.src_path)
+        self.app.update_repos_list()
+
+    def on_deleted(self, event):
+        if type(event) != DirDeletedEvent:
+            return
+        matching = filter(lambda existing_repo : existing_repo.path == event.src_path, self.app.repos)
+
+        map(lambda x: self.app.repos.remove(x), matching)
+        logger.info("%s was deleted; removing from klaus", event.src_path)
+        self.app.update_repos_list()
+
+    def on_moved(self, event):
+        """
+        Moved is not the same as deleted, unless the moved dir is no longer a subset of
+        any of the watched trees. In that case we need to consider it a deletion,
+        and remove it from the list of watched repos; otherwise we need to update its path.
+        This is a potentially expensive operation, because the only reliable way
+        to detect whether it is a subset (including symlinks etc) is to walk the path.
+        In practice, this shouldn't affect performance in the general case,
+        as moving a repository is a rare event.
+        """
+
+        matching = filter(lambda existing_repo : existing_repo.path == event.src_path, self.app.repos)
+        if not matching:
+            logger.debug("ignored non-repo move, %s -> %s", event.src_path, event.dest_path)
+            return
+
+        #first unregister old repositories
+        map(lambda x: self.app.repos.remove(x), matching)
+        # then see if the new destination is part of the klaus watchlist
+        for directory in self.app.basedirs:
+            if is_subdirectory(directory, event.dest_path):
+                logger.info("'%s' -> '%s'; updating references", event.src_path, event.dest_path)
+                self._add_dir(event.dest_path)
+
+        self.app.update_repos_list()
+        logger.info(
+            "%s was moved out of KLAUS_REPOS to %s, and is no longer a watched repository",
+            event.src_path,
+            event.dest_path
+        )
+
+
+class KlausAutoDiscover(Klaus):
```

Comment:
This whole initialization business doesn't even need a subclass since none of the internal logic of the `Klaus` class is changed. I think it's more elegant to just use the normal `Klaus` instance and do the watchdog stuff in a bootstrapping function.